Enforce typed error structures for better debugging and dashboard display

## Problem

When TypeScript workflows fail in the Mediar platform, error messages are inconsistently formatted and often buried in nested JSON structures. This makes debugging difficult and results in poor UX in the dashboard.

### Current Behavior

**Example of a workflow error:**
```json
{
  "stdout": "{\"result\":{\"error\":\"Cannot find module './steps/00-sign-out'\"}}"
}
```

**What users see in the dashboard:**
```
❌ Failed to execute tool: execute_sequence
```

**What they should see:**
```
❌ Cannot find module './steps/00-sign-out' from 'C:\...\terminator.ts'
```

### Root Cause

1. **No enforced error structure**: Workflows can return errors in any format
2. **Nested JSON parsing required**: Errors are wrapped in stdout/stderr strings
3. **Generic error wrapping**: MCP layer adds generic messages that hide the real error
4. **Dashboard can't parse errors**: No consistent structure for the UI to extract error details

## Proposed Solution

### 1. Define Typed Error Structures

Add typed error returns to the workflow runtime:

```typescript
// packages/workflow-runtime/src/types.ts

export type WorkflowError = {
  /** Error category for better filtering/alerting */
  category: 'user_error' | 'system_error' | 'network_error' | 'timeout' | 'validation_error';
  
  /** Human-readable error message */
  message: string;
  
  /** Unique error code (e.g., 'MODULE_NOT_FOUND', 'SELECTOR_NOT_FOUND') */
  code: string;
  
  /** Optional context for debugging */
  context?: {
    file?: string;
    line?: number;
    step?: string;
    moduleName?: string;
    selector?: string;
    [key: string]: any;
  };
  
  /** Original error stack trace */
  stack?: string;
};

export type WorkflowResult<T = any> = 
  | { 
      success: true; 
      data: T;
      metadata?: {
        executionTime?: number;
        stepsCompleted?: number;
      };
    }
  | { 
      success: false; 
      error: WorkflowError;
    };
```

### 2. Runtime Error Wrapping

Update the workflow runtime to catch and wrap all errors:

```typescript
// packages/workflow-runtime/src/executor.ts

export async function executeWorkflow<T>(
  workflow: WorkflowDefinition
): Promise<WorkflowResult<T>> {
  try {
    const result = await runWorkflow(workflow);
    return {
      success: true,
      data: result,
      metadata: {
        executionTime: Date.now() - startTime,
        stepsCompleted: completedSteps
      }
    };
  } catch (error) {
    return {
      success: false,
      error: normalizeError(error)
    };
  }
}

function normalizeError(error: unknown): WorkflowError {
  if (error instanceof ModuleNotFoundError) {
    return {
      category: 'user_error',
      message: `Cannot find module '${error.moduleName}'`,
      code: 'MODULE_NOT_FOUND',
      context: {
        moduleName: error.moduleName,
        file: error.file,
        line: error.line
      },
      stack: error.stack
    };
  }
  
  if (error instanceof SelectorNotFoundError) {
    return {
      category: 'user_error',
      message: `Element not found: ${error.selector}`,
      code: 'SELECTOR_NOT_FOUND',
      context: {
        selector: error.selector,
        step: error.stepName,
        screenshot: error.screenshotUrl
      },
      stack: error.stack
    };
  }
  
  // Generic error fallback
  return {
    category: 'system_error',
    message: error instanceof Error ? error.message : String(error),
    code: 'UNKNOWN_ERROR',
    stack: error instanceof Error ? error.stack : undefined
  };
}
```

### 3. MCP Layer Changes

Update the MCP `execute_sequence` tool to return structured errors directly:

```typescript
// Instead of wrapping in stdout/stderr:
return {
  error: {
    category: 'user_error',
    message: 'Cannot find module',
    code: 'MODULE_NOT_FOUND',
    context: { ... }
  }
};

// Not this:
return {
  stdout: JSON.stringify({ error: '...' })
};
```

### 4. Dashboard Integration

With structured errors, the dashboard can:

```typescript
// Frontend can now do:
if (!result.success) {
  const { error } = result;
  
  // Show user-friendly message
  showError(error.message);
  
  // Filter by category
  if (error.category === 'user_error') {
    showFixSuggestions(error.code);
  }
  
  // Display context
  if (error.context?.file) {
    highlightErrorInEditor(error.context.file, error.context.line);
  }
}
```

## Benefits

1. **Better DX**: Developers see exact errors immediately
2. **Better UX**: Dashboard can show helpful suggestions based on error codes
3. **Better monitoring**: Can alert/filter by error category
4. **Better debugging**: Structured context makes root cause analysis easier
5. **Backward compatible**: Can wrap legacy errors in the new structure

## Implementation Plan

1. [ ] Add `WorkflowError` and `WorkflowResult` types to workflow runtime
2. [ ] Create `normalizeError()` helper for common error types
3. [ ] Update workflow executor to wrap all errors
4. [ ] Update MCP layer to return structured errors
5. [ ] Update rust-executor to parse new error format
6. [ ] Update dashboard to display structured errors
7. [ ] Add migration guide for existing workflows

## Alternative Considered

**Status Code Approach**: Use HTTP-style status codes (404 for not found, 500 for internal error)
- ❌ Less descriptive than category + code
- ❌ Limited to numeric codes
- ✅ Familiar pattern for developers

**Current Workaround**: The rust-executor now parses error chains and extracts details from JSON, but this is fragile and doesn't solve the root issue of inconsistent error formats.

## Related Issues

- #369 (DX improvements)
- https://github.com/mediar-ai/mediar-web-app/issues (dashboard error display)

## Priority

**High** - This affects every workflow error and makes debugging significantly harder for users.

--- a/packages/workflow/src/index.ts
+++ b/packages/workflow/src/index.ts
@@ -8,7 +8,7 @@
 export { createStep } from './step';
 export { createWorkflow, WorkflowBuilder } from './workflow';
 export { createWorkflowRunner, WorkflowRunner } from './runner';
-export { createWorkflowError } from './types';
+export { WorkflowError, createWorkflowError } from './types';
 
 export type {
   Desktop,

--- a/packages/workflow/src/types.ts
+++ b/packages/workflow/src/types.ts
@@ -134,7 +134,7 @@ export interface ExecuteError {
  *
  * @example
  * ```typescript
- * throw createWorkflowError({
+ * throw WorkflowError({
  *   category: 'business',
  *   code: 'SAP_DUPLICATE_INVOICE',
  *   message: 'Invoice already exists in SAP',
@@ -143,7 +143,7 @@ export interface ExecuteError {
  * });
  * ```
  */
-export function createWorkflowError(error: ExecuteError): Error & ExecuteError {
+export function WorkflowError(error: ExecuteError): Error & ExecuteError {
   const err = new Error(error.message) as Error & ExecuteError;
   err.category = error.category;
   err.code = error.code;
@@ -151,6 +151,11 @@ export function createWorkflowError(error: ExecuteError): Error & ExecuteError {
   err.metadata = error.metadata;
   return err;
 }
+
+/**
+ * @deprecated Use WorkflowError instead (follows Error/TypeError naming convention)
+ */
+export const createWorkflowError = WorkflowError;
 
 /**
  * Workflow execution response

